rfcs/0000-structured-logging.md

```diff
@@ -0,0 +1,2189 @@
+
+        #[cfg(not(feature = "kv_serde"))]
+        impl Visit for char {}
+        #[cfg(not(feature = "kv_serde"))]
+        impl Visit for bool {}
+
+        #[cfg(not(feature = "kv_serde"))]
+        impl<T> Visit for Option<T>
+        where
+            T: Visit {}
+
+        #[cfg(all(not(feature = "kv_serde"), feature = "std"))]
+        impl<T: ?Sized> Visit for Box<T>
+        where
+            T: Visit {}
+
+        #[cfg(not(feature = "kv_serde"))]
+        impl<'a> Visit for &'a str {}
+        #[cfg(all(not(feature = "kv_serde"), feature = "std"))]
+        impl Visit for String {}
+
+        #[cfg(not(feature = "kv_serde"))]
+        impl<'a> Visit for &'a [u8] {}
+        #[cfg(all(not(feature = "kv_serde"), feature = "std"))]
+        impl Visit for Vec<u8> {}
+
+        #[cfg(not(feature = "kv_serde"))]
+        impl<'a, T> Visit for &'a T
+        where
+            T: Visit {}
+
+        #[cfg(feature = "kv_serde")]
+        impl<T> Visit for T
+        where
+            T: Debug + Serialize {}
+    }
+
+    pub use source::Source;
+
+    /// An error encountered while visiting key-value pairs.
+    pub struct Error {}
+
+    impl Error {
+        /// Create an error from a static message.
+        pub fn msg(msg: &'static str) -> Self {}
+
+        /// Get a reference to a standard error.
+        #[cfg(feature = "std")]
+        pub fn as_error(&self) -> &(dyn std::error::Error + Send + Sync + 'static) {}
+
+        /// Convert into a standard error.
+        #[cfg(feature = "std")]
+        pub fn into_error(self) -> Box<dyn std::error::Error + Send + Sync> {}
+
+        /// Convert into a `serde` error.
+        #[cfg(feature = "kv_serde")]
+        pub fn into_serde<E>(self) -> E
+        where
+            E: serde::ser::Error {}
+    }
+
+    #[cfg(not(feature = "std"))]
+    impl From<std::fmt::Error> for Error {}
+
+    #[cfg(feature = "std")]
+    impl<E> From<E> for Error
+    where
+        E: std::error::Error {}
+
+    #[cfg(feature = "std")]
+    impl From<Error> for Box<dyn std::error::Error + Send + Sync> {}
+
+    #[cfg(feature = "std")]
+    impl AsRef<dyn std::error::Error + Send + Sync + 'static> for Error {}
+}
+```
+
+### `Record` and `RecordBuilder`
+
+Structured key-value pairs can be set on a `RecordBuilder`:
+
+```rust
+impl<'a> RecordBuilder<'a> {
+    /// Set key values
+    pub fn key_values(&mut self, kvs: ErasedSource<'a>) -> &mut RecordBuilder<'a> {
+        self.record.kvs = kvs;
+        self
+    }
+}
+```
+
+These key-value pairs can then be accessed on the built `Record`:
+
+```rust
+#[derive(Clone, Debug)]
+pub struct Record<'a> {
+    ...
+
+    kvs: ErasedSource<'a>,
+}
+
+impl<'a> Record<'a> {
+    /// The key value pairs attached to this record.
+    /// 
+    /// Pairs aren't guaranteed to be unique (the same key may be repeated with different values).
+    pub fn key_values(&self) -> ErasedSource {
+        self.kvs.clone()
+    }
+}
+```
+
+### `Error`
```

Comment:
Do we really need a custom error? Is there any serialisation format that can fail? Can't this be relaced by `io::Error`?

Reply:
Json can fail if you give it non-string keys, and any printing done through the `Write` trait can also fail. We do need our own error type because the `log` crate needs to work in no-std environments, and `io::Error` is only available in the standard library.